ldap/servers/plugins/referint/referint.c

```
@@ -726,6 +726,22 @@ _do_modify(Slapi_PBlock *mod_pb, Slapi_DN *entrySDN, LDAPMod **mods)
     slapi_modify_internal_pb(mod_pb);
     slapi_pblock_get(mod_pb, SLAPI_PLUGIN_INTOP_RESULT, &rc);
 
+    /* Do we need to override the return value */
+    if (rc) {
+        for (size_t i = 0; mods && mods[i] != NULL; i++) {
+            int mod_type = mods[i]->mod_op & LDAP_MOD_OP;
+            if (mod_type == LDAP_MOD_DELETE) {
+                if (rc == LDAP_NO_SUCH_ATTRIBUTE) {
+                    rc = LDAP_SUCCESS;
```

This is changing the 'rc' value, so why not breaking from the loop ?
I have a mix feeling about the loop. If there is
MOD_DEL -> LDAP_SUCCESS
MOD_DEL -> NO_SUCH_ATTR
MOD_ADD-> OPERATION_ERROR
It will hide the OPERATION_ERROR.

---

This function can be invoked for either "delete" mode where there is a single MOD_DEL, or "modrdn" mode with MOD_DEL and MOD_ADD operations. So IIUC, i don't think the scenario you described can happen. 

Regarding why I didn't break out of the loop:
In "delete" mode a single MOD_DEL is used and the loop will exit anyway. 
In "modrdn" mode where there are 2 mods, MOD_DEL and MOD_ADD, in a worse case scenario where both mods fail the pblock intop result will be set to the MOD_ADD error code. When this is detected, the loop will be on its final iteration. 

But I can add breaks if you prefer.

---

Anyway the loop logic is broken:
   if the internal modify fails no change is done and you will return success without adding/removing the other values that must be added/deleted.
   
When we look at _do_modify I see it ts called in two ways:
   [1] with a single modifier (in that case, we should use a _do_single_modify function that calls _do_modify and returns LDAP_SUCCES if return code is LDAP_NO_SUCH_ATTRIBUTE or  LDAP_TYPE_OR_VALUE_EXISTS
   [2] with a list of modifier. The fact that the value is already present/deleted is already tested when building the modifier list means that there is nothing else to do. just go on to call do_modifier in that case.

---

> Anyway the loop logic is broken:
> if the internal modify fails no change is done and you will return success without adding/removing the other values that must be added/deleted.

In the context of updating group member attribute values that can also be updated by AM and MO plugins, it is my understanding that our goal here is to gracefully handle a failed modify with these mod op -> error code characteristics, any other mod op error should be returned to the caller:
MOD_DEL -> LDAP_NO_SUCH_ATTRIBUTE
MOD_ADD -> LDAP_TYPE_OR_VALUE_EXISTS

---

I think that the term "gracefully handle" is a bit unclear:
  - it means the the plugin should not fail and abort the operation in that case
    and I guess we agree on that point because your commit achieve that goal.
  - but it should not either left thing in inconsistent state and aborts all the plugin changes because one of them is failing.

---

BTW: I think we have the same problem with the memberof_add_memberof_attr change.

I wonder if the easiest solution is not to perform the internal modify as it is today 
 and if it fails with LDAP_NO_SUCH_ATTRIBUTE/LDAP_TYPE_OR_VALUE_EXISTS:
   if there is only 1 modifier:  do return success
   otherwise performs one internal modify operation  per modifier ignoring   LDAP_NO_SUCH_ATTRIBUTE/LDAP_TYPE_OR_VALUE_EXISTS error. 
   
   Anyway the error case is quite rare so the perf impact should stay reasonable,
     and the change in both plugins could be simple if a new slapi_special_internal_modify operation is added in libslapd

---

Another way it to add an new operation flags and test it in modify operation to ignores the errors ...
(That will probably to more efficient)

---

I initially had a performance concern with the idea of one internal_modify by  MOD (e.g. MOD_DEL, MOD_ADD). At a second though I agree with @progier389 that the performance hit is rare and limited. MODRDN is not that frequent and limited because the number of groups that contain a given entry is usually limited. So having two MOD rather than one looks okay.
If MOD_DEL fails because of LDAP_NO_SUCH_ATTRIBUTE, just ignore the error, if MOD_ADD fails because of LDAP_TYPE_OR_VALUE_EXISTS just ignore the error. This is more expensive but makes the code easier to maintain.